### ISSUE TITLE
Some kind of "segmented" Buf

### ISSUE THREAD

**Issue by vorner** [tokio-rs/bytes#462]
Hello

I'm thinking about the following use case. Data come through network (possibly a body of http POST from hyper) in chunks. Each chunk is `Bytes`. I want to consume them all and then parse somehow, maybe with prost. Prost is happy to take a `Buf`.

Right now I still need to append everything to a Vec o BytesMut by copying things. Would it make sense to have some kind of segmented `Buf` implementation, that can add more segments? Something like `Segmented(Vec<impl Buf>)` or `Segmented(Iterator<Item: Buf>)`? Something a bit like `Chain`, but with possibility to construct it with more than 2 parts.

If there would be interest in something like that, I can send a PR. But I would like to check first, before I jump to implementing it.

**Comment by Darksonn** [comment]
I think it is reasonable to have an analogue to `Chain` with a variable number of elements. Note that the hyper crate provides this kind of utility as [`aggregate`][1].

[1]: https://docs.rs/hyper/0.14/hyper/body/fn.aggregate.html

**Comment by vorner** [comment]
Thanks for the suggestion. Unfortunately, that one:

* Returns `impl Buf`, not a named type, which makes it annoying to work with.
* Is usable only with hyper, not with general things.
* Even with hyper, one doesn't have the needed flexibility. For example, it's not possible to implement some kind of huge body protection ‒ checking the size can be done only after the whole body is sitting in RAM, while I'd like to do such thing after each chunk.

OK, I'll try to do the PR and see if it is simple enough.

**Comment by Darksonn** [comment]
Sure, it was mostly to point at similar things in the ecosystem.

**Comment by seanmonstar** [comment]
I've brought this up a couple times, it would generally be useful. However, it might be better to first appear in something like `tokio-util`.

**Comment by Ralith** [comment]
`Bytes` often appears in the public APIs of crates, so breaking it is very disruptive; anything that can be in a separate crate probably should be.

**Comment by carllerche** [comment]
I will echo the sentiment that a segmented buffer would be useful but is out of scope for the `bytes` crate. Eventually, we could possibly have a separate crate w/ a collection of various buffer types, but for now, I don't think anyone has the capacity to take it on as part of Tokio.

I'm going to close the issue as there is no further action to take here.

**Comment by vorner** [comment]
So, does anyone have an idea of a place to offer it? I'm not sure this is big enough to warrant me starting a brand new crate (though I could certainly do that too; one more crate with not enough time to give it all the love it wants won't be a big difference).

**Comment by carllerche** [comment]
I don't think there is anything wrong w/ small crates, at least to start.

**Comment by vorner** [comment]
No, it's not exactly „wrong“. The discoverability and bus factor are a bit worse for them, though, so I always prefer to „donate“ the code to bigger and well established ones first, if possible.

If/once I create such crate, would the bytes crate or some tokio tutorial be interested in linking/show-casing it as some kind of third-party utility?

**Comment by vorner** [comment]
For whoever finds this issue in the future, I've written the crate and published it: https://crates.io/crates/bytes-utils, the interesting types are [SegmentedBuf](https://docs.rs/bytes-utils/0.1.1/bytes_utils/struct.SegmentedBuf.html), [SegmentedSlice](https://docs.rs/bytes-utils/0.1.1/bytes_utils/struct.SegmentedSlice.html). There's also an accompanying blog post with the motivation and what I want to use it for: https://vorner.github.io/2021/01/31/saving-some-allocations.html.
